### PR TITLE
adding variant to kokkos to enable building OneAPI device SPIR

### PR DESCRIPTION
**Pull request by frankwillmore**
stubbing out space to add flags to enable building device spir (intermediate representation)

### CODE REVIEW

**Comment by frankwillmore** [comment]
@jjwilke @janciesko 

I'm trying to put together a patch to add a variant spir that will add the compiler option -fopenmp-targets=spir64 when building with oneapi. Without out this option, the oneapi compilers will just build host code.  I defer to maintainers on how best to get this inserted. The other option -fiopenmp is more general and I will address in the compiler class in a separate pull request.

**Comment by haampie** [comment]
```

var/spack/repos/builtin/packages/kokkos/package.py:282: [E115] expected an indented block (comment)
var/spack/repos/builtin/packages/kokkos/package.py:283: [E115] expected an indented block (comment
```

**Comment by frankwillmore** [comment]
> ```
> 
> var/spack/repos/builtin/packages/kokkos/package.py:282: [E115] expected an indented block (comment)
> var/spack/repos/builtin/packages/kokkos/package.py:283: [E115] expected an indented block (comment
> ```

I made the entire block here a comment, as this is a (non-working) prototype. 

My intent here is to raise awareness with the developers, as an additional compiler flag needs to be added via the build system. There was a variant 'sycl' added to the package, which has no associated action. I propose the spir variant for oneapi-specific builds. The software is not ready for oneapi deployment unless  ```-fopenmp-targets=spir64``` or equivalent is supported. I leave it to the package maintainers to decide how oneapi gets this flag.

**Comment by crtrott** [comment]
I think this should be added automatically by Kokkos if setting the correct Arch in addition to SYCL backend or?

**Comment by DavidPoliakoff** [comment]
Yeah, we try to avoid having just direct flags passed on to the compiler, things tend to evolve over time. The right way to do this is likely to add an Arch to Kokkos itself, then expose that through the Spack package. Let me know if you're interested in pursuing that and I can walk you through it

**Comment by DavidPoliakoff** [comment]
Ah, looks like the arch exists, it's called KOKKOS_ARCH_INTEL_GEN. Lemme check if we expose that in Spack. edit: nope, don't think we do

**Comment by masterleinad** [comment]
> stubbing out space to add flags to enable building device spir (intermediate representation)

You can totally just enable the `SYCL` backend and no extra flags are necessary (using JIT compilation).

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (crtrott) [inline var/spack/repos/builtin/packages/kokkos/package.py:38]: Note this is not enough and also not the right thing to do for Kokkos. We need to add openmptarget device, and IntelGEN architecture.

**Review by crtrott — CHANGES_REQUESTED** [review]
This wouldn't do the right thing for Kokkos and is not the right way to fix this.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (frankwillmore) [reply var/spack/repos/builtin/packages/kokkos/package.py:38]: Yes, this is just a strawman. I have no stake in how this implemented so long as the correct backend is generated and the spack package knows how to communicate that information to cmake. 

Is the intended use of the 'sycl' variant in the package for oneapi, a different backend, or generally? It's never referenced in the package.

Comment (crtrott) [reply var/spack/repos/builtin/packages/kokkos/package.py:38]: sycl is a different backend. To target Intel GPUs with OneAPI you can either use their OpenMP Device compiler or the DPC++/SYCL programming model. The latter does not need that spir option, the OpenMPTarget backend does, its in our cmake but not fully exposed in SPACK. Is there an exisitng spack architecture option for Intel GPUs?

Comment (frankwillmore) [reply var/spack/repos/builtin/packages/kokkos/package.py:38]: @becker33 please correct me, but AFAIK spack doesn't track device architecture. 

The SPIR code in my understanding is device independent and JIT-ed, so device arch shouldn't matter for the openmp case, but would thus matter for the SYCL backend? @TApplencourt does this sound right?

Comment (masterleinad) [reply var/spack/repos/builtin/packages/kokkos/package.py:38]: The device architecture isn't required either `SYCL` either. Without extra flags, you get `JIT` compilation but specifying `KOKKOS_ARCH_INTEL_GEN` enables `AOT` compilation for `Intel Gen9` GPUs.
Since we need extra flags for `OpenMPTarget` to produce `SPIR` code (as opposed to targeting other backends like `CUDA`) and we are mostly interested in `SPIR` code for `Intel GPUs`, we decided to also use `KOKKOS_ARCH_INTEL_GEN` to add the correct flags.

**Comment by alalazo** [comment]
Is this still a work in progress?

**Comment by frankwillmore** [comment]
I am no longer involved in this effort and defer to kokkos maintainers for any outstanding issues.

**Pull request closed without merging**
